docs/markdown/Writing Plugins/tutorials/01-create-new-goal-with-a-custom-target.md

```diff
@@ -0,0 +1,335 @@
+
+We could [add a custom field](https://www.pantsbuild.org/docs/target-api-new-fields) to provide a file path, however, when using the `source` field, you get a few features for free such as setting the `default` value and `expected_file_extensions`. Furthermore, with the `source` field, thanks to the [`unmatched_build_file_globs`](https://www.pantsbuild.org/docs/reference-global#unmatched_build_file_globs) option, you won't need to provide custom logic to handle errors when a path glob doesn't expand to any files in your repository.
+
+You only need to register the target to start using it: 
+
+```python pants-plugins/project_version/register.py
+from project_version.targets import ProjectVersionTarget
+
+
+def target_types():
+    return [ProjectVersionTarget]
+```
+
+You can now run `./pants help version_file` to learn more about the target. You can now also add a target to a `myapp/BUILD` file:
+
+```python
+version_file(
+    name="main-project-version", 
+    source="VERSION",
+)
+```
+
+Since you have registered the target, Pants will be able to "understand" it:
+
+```text
+$ ./pants peek myapp:main-project-version
+[
+  {
+    "address": "myapp:main-project-version",
+    "target_type": "version_file",
+    "dependencies": [],
+    "description": null,
+    "source_raw": "VERSION",
+    "sources": [
+      "myapp/VERSION"
+    ],
+    "tags": null
+  }
+]
+```
+
+## Creating a goal
+
+[Goals](https://www.pantsbuild.org/docs/goals) are the commands that Pants runs such as `fmt` or `lint`. To be able to get a project version number (using a target we've just created), we need to [create a new goal](https://www.pantsbuild.org/docs/rules-api-goal-rules). The code below is the boilerplate necessary to create a goal, so it's not really necessary to understand how, for instance, subsystems work right now. The function decorated with the `@goal_rule` can be named anything, but it's helpful for the name to represent the functionality your goal provides. To make your goal part of the plugin's interface, add it to the `rules` function in the `register.py` module.
+
+```python pants-plugins/project_version/rules.py
+from pants.engine.goal import Goal, GoalSubsystem
+from pants.engine.rules import collect_rules, goal_rule
+
+
+class ProjectVersionSubsystem(GoalSubsystem):
+    name = "project-version"
+    help = "Show representation of the project version from the `VERSION` file."
+
+
+class ProjectVersionGoal(Goal):
+    subsystem_cls = ProjectVersionSubsystem
+
+
+@goal_rule
+async def goal_show_project_version() -> ProjectVersionGoal:
+    return ProjectVersionGoal(exit_code=0)
+
+
+def rules():
+    return collect_rules()
+```
+```python pants-plugins/project_version/rules.py
+from project_version.targets import ProjectVersionTarget
+import project_version.rules as project_version_rules
+
+
+def target_types():
+    return [ProjectVersionTarget]
+
+
+def rules():
```

Comment:
` -> Iterable[Rule | UnionRule]:`

Reply:
I think it would be better to offer a type alias we export from Pants that could be used here, that way plugin authors won't have to adapt to changes when upgrading, besides fixing actual issues. Also since I think this one is tricky.. as there's potentially `QueryRule`s too.. and maybe others in the future?

Reply:
👍🏽 So long as it's typed, I'm on board :)

Reply:
Okay, I am not sure which one you folks would recommend to use if at all :confused: